src/aiohomeconnect/client.py

```diff
@@ -69,6 +77,25 @@ async def request(self, method: str, url: str, **kwargs: Any) -> Response:
             headers=headers,
         )
 
+    @asynccontextmanager
+    async def connect_sse(
+        self,
+        method: str,
+        url: str,
+        **kwargs: Any,
+    ) -> AsyncIterator[EventSource]:
+        """Create a SSE connection."""
+        headers = await self._get_headers(kwargs.pop("headers", None))
```

Comment:
Something to think about for the future, which we don't need to solve in this PR: We may need to think of a solution for how to check for an expired access token, if this connection stays active longer than the token expiration time. I'm not sure what the API behavior is, but with the library Home Assistant is using at the moment this is an issue, I believe. I think there's an error response when the token is expired. I'll check next time I get it.

The access token is expiration time is 24 hours according to the API docs.

Reply:
I just prepared an experiment to see what will happen in 24h after token expiration, I tried to emulate an error with the simulator, but the stream doesn't break. So I ran two instances, one for the simulator and another for my own Home Connect, to see what will happen in 24h.

Reply:
Okey, both experiments ended before 24h
Simulator APIraised this exception:
```
RemoteProtocolError: peer closed connection without sending complete message body (incomplete chunked read)
```
(see more details on traceback from below)

Real API died in complete silence, last event was 8 hours ago, but I should have printed KEEP-ALIVE events together with a timelog, I will repeat the experiment to get more data

<details>
<summary>Traceback</summary>

```
╭───────────────────── Traceback (most recent call last) ──────────────────────╮
│ /root/.cache/pypoetry/virtualenvs/aiohomeconnect-oJCSMThX-py3.11/lib/python3 │
│ .11/site-packages/httpx/_transports/default.py:101 in                        │
│ map_httpcore_exceptions                                                      │
│                                                                              │
│    98 │   if len(HTTPCORE_EXC_MAP) == 0:                                     │
│    99 │   │   HTTPCORE_EXC_MAP = _load_httpcore_exceptions()                 │
│   100 │   try:                                                               │
│ ❱ 101 │   │   yield                                                          │
│   102 │   except Exception as exc:                                           │
│   103 │   │   mapped_exc = None                                              │
│   104                                                                        │
│                                                                              │
│ ╭───────────────────────────────── locals ─────────────────────────────────╮ │
│ │ message = 'peer closed connection without sending complete message body  │ │
│ │           (incomplete chunked'+6                                         │ │
│ ╰──────────────────────────────────────────────────────────────────────────╯ │
│                                                                              │
│ /root/.cache/pypoetry/virtualenvs/aiohomeconnect-oJCSMThX-py3.11/lib/python3 │
│ .11/site-packages/httpx/_transports/default.py:271 in __aiter__              │
│                                                                              │
│   268 │                                                                      │
│   269 │   async def __aiter__(self) -> typing.AsyncIterator[bytes]:          │
│   270 │   │   with map_httpcore_exceptions():                                │
│ ❱ 271 │   │   │   async for part in self._httpcore_stream:                   │
│   272 │   │   │   │   yield part                                             │
│   273 │                                                                      │
│   274 │   async def aclose(self) -> None:                                    │
│                                                                              │
│ ╭───────────────────────────────── locals ─────────────────────────────────╮ │
│ │ part = b'event: KEEP-ALIVE\ndata: ""\n\n\n'                              │ │
│ │ self = <httpx._transports.default.AsyncResponseStream object at          │ │
│ │        0x7f88d4fa90>                                                     │ │
│ ╰──────────────────────────────────────────────────────────────────────────╯ │
│                                                                              │
│ /root/.cache/pypoetry/virtualenvs/aiohomeconnect-oJCSMThX-py3.11/lib/python3 │
│ .11/site-packages/httpcore/_async/connection_pool.py:367 in __aiter__        │
│                                                                              │
│   364 │   │   │   │   yield part                                             │
│   365 │   │   except BaseException as exc:                                   │
│   366 │   │   │   await self.aclose()                                        │
│ ❱ 367 │   │   │   raise exc from None                                        │
│   368 │                                                                      │
│   369 │   async def aclose(self) -> None:                                    │
│   370 │   │   if not self._closed:                                           │
│                                                                              │
│ ╭───────────────────────────────── locals ─────────────────────────────────╮ │
│ │ part = b'event: KEEP-ALIVE\ndata: ""\n\n\n'                              │ │
│ │ self = <httpcore._async.connection_pool.PoolByteStream object at         │ │
│ │        0x7f88e26ed0>                                                     │ │
│ ╰──────────────────────────────────────────────────────────────────────────╯ │
│                                                                              │
│ /root/.cache/pypoetry/virtualenvs/aiohomeconnect-oJCSMThX-py3.11/lib/python3 │
│ .11/site-packages/httpcore/_async/connection_pool.py:363 in __aiter__        │
│                                                                              │
│   360 │                                                                      │
│   361 │   async def __aiter__(self) -> AsyncIterator[bytes]:                 │
│   362 │   │   try:                                                           │
│ ❱ 363 │   │   │   async for part in self._stream:                            │
│   364 │   │   │   │   yield part                                             │
│   365 │   │   except BaseException as exc:                                   │
│   366 │   │   │   await self.aclose()                                        │
│                                                                              │
│ ╭───────────────────────────────── locals ─────────────────────────────────╮ │
│ │ part = b'event: KEEP-ALIVE\ndata: ""\n\n\n'                              │ │
│ │ self = <httpcore._async.connection_pool.PoolByteStream object at         │ │
│ │        0x7f88e26ed0>                                                     │ │
│ ╰──────────────────────────────────────────────────────────────────────────╯ │
│                                                                              │
│ /root/.cache/pypoetry/virtualenvs/aiohomeconnect-oJCSMThX-py3.11/lib/python3 │
│ .11/site-packages/httpcore/_async/http11.py:349 in __aiter__                 │
│                                                                              │
│   346 │   │   │   # before raising that exception.                           │
│   347 │   │   │   with AsyncShieldCancellation():                            │
│   348 │   │   │   │   await self.aclose()                                    │
│ ❱ 349 │   │   │   raise exc                                                  │
│   350 │                                                                      │
│   351 │   async def aclose(self) -> None:                                    │
│   352 │   │   if not self._closed:                                           │
│                                                                              │
│ ╭───────────────────────────────── locals ─────────────────────────────────╮ │
│ │  chunk = b'event: KEEP-ALIVE\ndata: ""\n\n\n'                            │ │
│ │ kwargs = {'request': <Request [b'GET']>}                                 │ │
│ │   self = <httpcore._async.http11.HTTP11ConnectionByteStream object at    │ │
│ │          0x7f88d87b10>                                                   │ │
│ ╰──────────────────────────────────────────────────────────────────────────╯ │
│                                                                              │
│ /root/.cache/pypoetry/virtualenvs/aiohomeconnect-oJCSMThX-py3.11/lib/python3 │
│ .11/site-packages/httpcore/_async/http11.py:341 in __aiter__                 │
│                                                                              │
│   338 │   │   kwargs = {"request": self._request}                            │
│   339 │   │   try:                                                           │
│   340 │   │   │   async with Trace("receive_response_body", logger, self._re │
│ ❱ 341 │   │   │   │   async for chunk in self._connection._receive_response_ │
│   342 │   │   │   │   │   yield chunk                                        │
│   343 │   │   except BaseException as exc:                                   │
│   344 │   │   │   # If we get an exception while streaming the response,     │
│                                                                              │
│ ╭───────────────────────────────── locals ─────────────────────────────────╮ │
│ │  chunk = b'event: KEEP-ALIVE\ndata: ""\n\n\n'                            │ │
│ │ kwargs = {'request': <Request [b'GET']>}                                 │ │
│ │   self = <httpcore._async.http11.HTTP11ConnectionByteStream object at    │ │
│ │          0x7f88d87b10>                                                   │ │
│ ╰──────────────────────────────────────────────────────────────────────────╯ │
│                                                                              │
│ /root/.cache/pypoetry/virtualenvs/aiohomeconnect-oJCSMThX-py3.11/lib/python3 │
│ .11/site-packages/httpcore/_async/http11.py:210 in _receive_response_body    │
│                                                                              │
│   207 │   │   timeout = timeouts.get("read", None)                           │
│   208 │   │                                                                  │
│   209 │   │   while True:                                                    │
│ ❱ 210 │   │   │   event = await self._receive_event(timeout=timeout)         │
│   211 │   │   │   if isinstance(event, h11.Data):                            │
│   212 │   │   │   │   yield bytes(event.data)                                │
│   213 │   │   │   elif isinstance(event, (h11.EndOfMessage, h11.PAUSED)):    │
│                                                                              │
│ ╭───────────────────────────────── locals ─────────────────────────────────╮ │
│ │    event = Data(                                                         │ │
│ │            │   data=bytearray(b'event: KEEP-ALIVE\ndata: ""\n\n\n'),     │ │
│ │            │   chunk_start=True,                                         │ │
│ │            │   chunk_end=True                                            │ │
│ │            )                                                             │ │
│ │  request = <Request [b'GET']>                                            │ │
│ │     self = <AsyncHTTP11Connection                                        │ │
│ │            ['https://simulator.home-connect.com:443', CLOSED, Request    │ │
│ │            Count: 1]>                                                    │ │
│ │  timeout = 60                                                            │ │
│ │ timeouts = {'connect': 60, 'read': 60, 'write': 60, 'pool': 60}          │ │
│ ╰──────────────────────────────────────────────────────────────────────────╯ │
│                                                                              │
│ /root/.cache/pypoetry/virtualenvs/aiohomeconnect-oJCSMThX-py3.11/lib/python3 │
│ .11/site-packages/httpcore/_async/http11.py:220 in _receive_event            │
│                                                                              │
│   217 │   │   self, timeout: Optional[float] = None                          │
│   218 │   ) -> Union[h11.Event, Type[h11.PAUSED]]:                           │
│   219 │   │   while True:                                                    │
│ ❱ 220 │   │   │   with map_exceptions({h11.RemoteProtocolError: RemoteProtoc │
│   221 │   │   │   │   event = self._h11_state.next_event()                   │
│   222 │   │   │                                                              │
│   223 │   │   │   if event is h11.NEED_DATA:                                 │
│                                                                              │
│ ╭───────────────────────────────── locals ─────────────────────────────────╮ │
│ │    data = b''                                                            │ │
│ │    self = <AsyncHTTP11Connection                                         │ │
│ │           ['https://simulator.home-connect.com:443', CLOSED, Request     │ │
│ │           Count: 1]>                                                     │ │
│ │ timeout = 60                                                             │ │
│ ╰──────────────────────────────────────────────────────────────────────────╯ │
│                                                                              │
│ /usr/lib/python3.11/contextlib.py:155 in __exit__                            │
│                                                                              │
│   152 │   │   │   │   # tell if we get the same exception back               │
│   153 │   │   │   │   value = typ()                                          │
│   154 │   │   │   try:                                                       │
│ ❱ 155 │   │   │   │   self.gen.throw(typ, value, traceback)                  │
│   156 │   │   │   except StopIteration as exc:                               │
│   157 │   │   │   │   # Suppress StopIteration *unless* it's the same except │
│   158 │   │   │   │   # was passed to throw().  This prevents a StopIteratio │
│                                                                              │
│ ╭───────────────────────────────── locals ─────────────────────────────────╮ │
│ │      self = <contextlib._GeneratorContextManager object at 0x7f88e84e90> │ │
│ │ traceback = <traceback object at 0x7f88db5180>                           │ │
│ │     value = RemoteProtocolError('peer closed connection without sending  │ │
│ │             complete message body (incomplete chunked read)')            │ │
│ ╰──────────────────────────────────────────────────────────────────────────╯ │
│                                                                              │
│ /root/.cache/pypoetry/virtualenvs/aiohomeconnect-oJCSMThX-py3.11/lib/python3 │
│ .11/site-packages/httpcore/_exceptions.py:14 in map_exceptions               │
│                                                                              │
│   11 │   except Exception as exc:  # noqa: PIE786                            │
│   12 │   │   for from_exc, to_exc in map.items():                            │
│   13 │   │   │   if isinstance(exc, from_exc):                               │
│ ❱ 14 │   │   │   │   raise to_exc(exc) from exc                              │
│   15 │   │   raise  # pragma: nocover                                        │
│   16                                                                         │
│   17                                                                         │
│                                                                              │
│ ╭───────────────────────────────── locals ─────────────────────────────────╮ │
│ │ map = {                                                                  │ │
│ │       │   <class 'h11._util.RemoteProtocolError'>: <class                │ │
│ │       'httpcore.RemoteProtocolError'>                                    │ │
│ │       }                                                                  │ │
│ ╰──────────────────────────────────────────────────────────────────────────╯ │
╰──────────────────────────────────────────────────────────────────────────────╯
RemoteProtocolError: peer closed connection without sending complete message 
body (incomplete chunked read)

The above exception was the direct cause of the following exception:

╭───────────────────── Traceback (most recent call last) ──────────────────────╮
│ /root/aiohomeconnect_simulator/src/aiohomeconnect/cli/__init__.py:102 in     │
│ subscribe_to_all_applicances_events                                          │
│                                                                              │
│    99 @cli.command()                                                         │
│   100 def subscribe_to_all_applicances_events(client_id: str, client_secret: │
│   101 │   """Subscribe and print events from all the appliances."""          │
│ ❱ 102 │   asyncio.run(_subscribe_to_all_applicances_events(client_id, client │
│   103                                                                        │
│   104                                                                        │
│   105 async def _subscribe_to_all_applicances_events(                        │
│                                                                              │
│ ╭───────────────────────────────── locals ─────────────────────────────────╮ │
│ │     client_id = '55B0EEF375D32F3F6BB6CEDA9CDE41C88361827B33D8DE2AD98330… │ │
│ │ client_secret = '043B5B6065A7F767FEBD396B0FC88EC883E22B041E1CD722726535… │ │
│ ╰──────────────────────────────────────────────────────────────────────────╯ │
│                                                                              │
│ /usr/lib/python3.11/asyncio/runners.py:190 in run                            │
│                                                                              │
│   187 │   │   │   "asyncio.run() cannot be called from a running event loop" │
│   188 │                                                                      │
│   189 │   with Runner(debug=debug) as runner:                                │
│ ❱ 190 │   │   return runner.run(main)                                        │
│   191                                                                        │
│   192                                                                        │
│   193 def _cancel_all_tasks(loop):                                           │
│                                                                              │
│ ╭───────────────────────────────── locals ─────────────────────────────────╮ │
│ │  debug = None                                                            │ │
│ │   main = <coroutine object _subscribe_to_all_applicances_events at       │ │
│ │          0x7f88f934c0>                                                   │ │
│ │ runner = <asyncio.runners.Runner object at 0x7f88ec8a10>                 │ │
│ ╰──────────────────────────────────────────────────────────────────────────╯ │
│                                                                              │
│ /usr/lib/python3.11/asyncio/runners.py:118 in run                            │
│                                                                              │
│   115 │   │                                                                  │
│   116 │   │   self._interrupt_count = 0                                      │
│   117 │   │   try:                                                           │
│ ❱ 118 │   │   │   return self._loop.run_until_complete(task)                 │
│   119 │   │   except exceptions.CancelledError:                              │
│   120 │   │   │   if self._interrupt_count > 0:                              │
│   121 │   │   │   │   uncancel = getattr(task, "uncancel", None)             │
│                                                                              │
│ ╭───────────────────────────────── locals ─────────────────────────────────╮ │
│ │        context = <_contextvars.Context object at 0x7f88ec9180>           │ │
│ │           coro = <coroutine object _subscribe_to_all_applicances_events  │ │
│ │                  at 0x7f88f934c0>                                        │ │
│ │           self = <asyncio.runners.Runner object at 0x7f88ec8a10>         │ │
│ │ sigint_handler = functools.partial(<bound method Runner._on_sigint of    │ │
│ │                  <asyncio.runners.Runner object at 0x7f88ec8a10>>,       │ │
│ │                  main_task=<Task finished name='Task-1'                  │ │
│ │                  coro=<_subscribe_to_all_applicances_events() done,      │ │
│ │                  defined at                                              │ │
│ │                  /root/aiohomeconnect_simulator/src/aiohomeconnect/cli/… │ │
│ │                  exception=RemoteProtocolError('peer closed connection   │ │
│ │                  without sending complete message body (incomplete       │ │
│ │                  chunked read)')>)                                       │ │
│ │           task = <Task finished name='Task-1'                            │ │
│ │                  coro=<_subscribe_to_all_applicances_events() done,      │ │
│ │                  defined at                                              │ │
│ │                  /root/aiohomeconnect_simulator/src/aiohomeconnect/cli/… │ │
│ │                  exception=RemoteProtocolError('peer closed connection   │ │
│ │                  without sending complete message body (incomplete       │ │
│ │                  chunked read)')>                                        │ │
│ ╰──────────────────────────────────────────────────────────────────────────╯ │
│                                                                              │
│ /usr/lib/python3.11/asyncio/base_events.py:653 in run_until_complete         │
│                                                                              │
│    650 │   │   if not future.done():                                         │
│    651 │   │   │   raise RuntimeError('Event loop stopped before Future comp │
│    652 │   │                                                                 │
│ ❱  653 │   │   return future.result()                                        │
│    654 │                                                                     │
│    655 │   def stop(self):                                                   │
│    656 │   │   """Stop running the event loop.                               │
│                                                                              │
│ ╭───────────────────────────────── locals ─────────────────────────────────╮ │
│ │   future = <Task finished name='Task-1'                                  │ │
│ │            coro=<_subscribe_to_all_applicances_events() done, defined at │ │
│ │            /root/aiohomeconnect_simulator/src/aiohomeconnect/cli/__init… │ │
│ │            exception=RemoteProtocolError('peer closed connection without │ │
│ │            sending complete message body (incomplete chunked read)')>    │ │
│ │ new_task = False                                                         │ │
│ │     self = <_UnixSelectorEventLoop running=False closed=True             │ │
│ │            debug=False>                                                  │ │
│ ╰──────────────────────────────────────────────────────────────────────────╯ │
│                                                                              │
│ /root/aiohomeconnect_simulator/src/aiohomeconnect/cli/__init__.py:110 in     │
│ _subscribe_to_all_applicances_events                                         │
│                                                                              │
│   107 ) -> None:                                                             │
│   108 │   """Subscribe and print events from all the appliances."""          │
│   109 │   client = CLIClient(client_id, client_secret)                       │
│ ❱ 110 │   async for event in client.stream_all_events():                     │
│   111 │   │   rich_print(event)                                              │
│   112                                                                        │
│   113                                                                        │
│                                                                              │
│ ╭───────────────────────────────── locals ─────────────────────────────────╮ │
│ │        client = <aiohomeconnect.cli.client.CLIClient object at           │ │
│ │                 0x7f88fea9d0>                                            │ │
│ │     client_id = '55B0EEF375D32F3F6BB6CEDA9CDE41C88361827B33D8DE2AD98330… │ │
│ │ client_secret = '043B5B6065A7F767FEBD396B0FC88EC883E22B041E1CD722726535… │ │
│ │         event = EventMessage(                                            │ │
│ │                 │   id='SIEMENS-HCS02DWH1-6BE58C26DCC1',                 │ │
│ │                 │   type=<EventType.STATUS: 'STATUS'>,                   │ │
│ │                 │   data=Event(                                          │ │
│ │                 │   │   key=<EventKey.BSH_COMMON_SETTING_POWER_STATE:    │ │
│ │                 'BSH.Common.Setting.PowerState'>,                        │ │
│ │                 │   │   timestamp=1733906986,                            │ │
│ │                 │   │   level='hint',                                    │ │
│ │                 │   │   handling='none',                                 │ │
│ │                 │   │   value='BSH.Common.EnumType.PowerState.On',       │ │
│ │                 │   │   ha_id=None,                                      │ │
│ │                 │   │   name=None,                                       │ │
│ │                 │   │                                                    │ │
│ │                 uri='/api/homeappliances/SIEMENS-HCS02DWH1-6BE58C26DCC1… │ │
│ │                 │   │   display_value=None,                              │ │
│ │                 │   │   unit=None                                        │ │
│ │                 │   )                                                    │ │
│ │                 )                                                        │ │
│ ╰──────────────────────────────────────────────────────────────────────────╯ │
│                                                                              │
│ /root/aiohomeconnect_simulator/src/aiohomeconnect/client.py:716 in           │
│ stream_all_events                                                            │
│                                                                              │
│   713 │   │   │   │   "Accept-Language": accept_language,                    │
│   714 │   │   │   },                                                         │
│   715 │   │   ) as event_source:                                             │
│ ❱ 716 │   │   │   async for sse in event_source.aiter_sse():                 │
│   717 │   │   │   │   if sse.event == EventType.KEEP_ALIVE:                  │
│   718 │   │   │   │   │   continue                                           │
│   719 │   │   │   │   if sse.event not in EventType._value2member_map_:      │
│                                                                              │
│ ╭───────────────────────────────── locals ─────────────────────────────────╮ │
│ │ accept_language = None                                                   │ │
│ │            data = {                                                      │ │
│ │                   │   'items': [                                         │ │
│ │                   │   │   {                                              │ │
│ │                   │   │   │   'level': 'hint',                           │ │
│ │                   │   │   │   'handling': 'none',                        │ │
│ │                   │   │   │   'key': 'BSH.Common.Setting.PowerState',    │ │
│ │                   │   │   │   'value':                                   │ │
│ │                   'BSH.Common.EnumType.PowerState.On',                   │ │
│ │                   │   │   │   'uri':                                     │ │
│ │                   '/api/homeappliances/SIEMENS-HCS02DWH1-6BE58C26DCC1/s… │ │
│ │                   │   │   │   'timestamp': 1733906986,                   │ │
│ │                   │   │   │   'haId': None                               │ │
│ │                   │   │   }                                              │ │
│ │                   │   ]                                                  │ │
│ │                   }                                                      │ │
│ │           event = Event(                                                 │ │
│ │                   │   key=<EventKey.BSH_COMMON_SETTING_POWER_STATE:      │ │
│ │                   'BSH.Common.Setting.PowerState'>,                      │ │
│ │                   │   timestamp=1733906986,                              │ │
│ │                   │   level='hint',                                      │ │
│ │                   │   handling='none',                                   │ │
│ │                   │   value='BSH.Common.EnumType.PowerState.On',         │ │
│ │                   │   ha_id=None,                                        │ │
│ │                   │   name=None,                                         │ │
│ │                   │                                                      │ │
│ │                   uri='/api/homeappliances/SIEMENS-HCS02DWH1-6BE58C26DC… │ │
│ │                   │   display_value=None,                                │ │
│ │                   │   unit=None                                          │ │
│ │                   )                                                      │ │
│ │    event_source = <httpx_sse._api.EventSource object at 0x7f8a85d750>    │ │
│ │            item = {                                                      │ │
│ │                   │   'level': 'hint',                                   │ │
│ │                   │   'handling': 'none',                                │ │
│ │                   │   'key': 'BSH.Common.Setting.PowerState',            │ │
│ │                   │   'value': 'BSH.Common.EnumType.PowerState.On',      │ │
│ │                   │   'uri':                                             │ │
│ │                   '/api/homeappliances/SIEMENS-HCS02DWH1-6BE58C26DCC1/s… │ │
│ │                   │   'timestamp': 1733906986,                           │ │
│ │                   │   'haId': None                                       │ │
│ │                   }                                                      │ │
│ │            self = <aiohomeconnect.cli.client.CLIClient object at         │ │
│ │                   0x7f88fea9d0>                                          │ │
│ │             sse = ServerSentEvent(event='message',                       │ │
│ │                   id='SIEMENS-HCS02DWH1-6BE58C26DCC1')                   │ │
│ ╰──────────────────────────────────────────────────────────────────────────╯ │
│                                                                              │
│ /root/.cache/pypoetry/virtualenvs/aiohomeconnect-oJCSMThX-py3.11/lib/python3 │
│ .11/site-packages/httpx_sse/_api.py:39 in aiter_sse                          │
│                                                                              │
│   36 │   async def aiter_sse(self) -> AsyncIterator[ServerSentEvent]:        │
│   37 │   │   self._check_content_type()                                      │
│   38 │   │   decoder = SSEDecoder()                                          │
│ ❱ 39 │   │   async for line in self._response.aiter_lines():                 │
│   40 │   │   │   line = line.rstrip("\n")                                    │
│   41 │   │   │   sse = decoder.decode(line)                                  │
│   42 │   │   │   if sse is not None:                                         │
│                                                                              │
│ ╭───────────────────────────────── locals ─────────────────────────────────╮ │
│ │ decoder = <httpx_sse._decoders.SSEDecoder object at 0x7f88d4e090>        │ │
│ │    line = ''                                                             │ │
│ │    self = <httpx_sse._api.EventSource object at 0x7f8a85d750>            │ │
│ │     sse = ServerSentEvent(event='message',                               │ │
│ │           id='SIEMENS-HCS02DWH1-6BE58C26DCC1')                           │ │
│ ╰──────────────────────────────────────────────────────────────────────────╯ │
│                                                                              │
│ /root/.cache/pypoetry/virtualenvs/aiohomeconnect-oJCSMThX-py3.11/lib/python3 │
│ .11/site-packages/httpx/_models.py:1031 in aiter_lines                       │
│                                                                              │
│   1028 │   async def aiter_lines(self) -> typing.AsyncIterator[str]:         │
│   1029 │   │   decoder = LineDecoder()                                       │
│   1030 │   │   with request_context(request=self._request):                  │
│ ❱ 1031 │   │   │   async for text in self.aiter_text():                      │
│   1032 │   │   │   │   for line in decoder.decode(text):                     │
│   1033 │   │   │   │   │   yield line                                        │
│   1034 │   │   │   for line in decoder.flush():                              │
│                                                                              │
│ ╭──────────────────────────── locals ────────────────────────────╮           │
│ │ decoder = <httpx._decoders.LineDecoder object at 0x7f8a93a6d0> │           │
│ │    line = ''                                                   │           │
│ │    self = <Response [200 OK]>                                  │           │
│ │    text = 'event: KEEP-ALIVE\ndata: ""\n\n\n'                  │           │
│ ╰────────────────────────────────────────────────────────────────╯           │
│                                                                              │
│ /root/.cache/pypoetry/virtualenvs/aiohomeconnect-oJCSMThX-py3.11/lib/python3 │
│ .11/site-packages/httpx/_models.py:1018 in aiter_text                        │
│                                                                              │
│   1015 │   │   decoder = TextDecoder(encoding=self.encoding or "utf-8")      │
│   1016 │   │   chunker = TextChunker(chunk_size=chunk_size)                  │
│   1017 │   │   with request_context(request=self._request):                  │
│ ❱ 1018 │   │   │   async for byte_content in self.aiter_bytes():             │
│   1019 │   │   │   │   text_content = decoder.decode(byte_content)           │
│   1020 │   │   │   │   for chunk in chunker.decode(text_content):            │
│   1021 │   │   │   │   │   yield chunk                                       │
│                                                                              │
│ ╭────────────────────────────── locals ───────────────────────────────╮      │
│ │ byte_content = b'event: KEEP-ALIVE\ndata: ""\n\n\n'                 │      │
│ │        chunk = 'event: KEEP-ALIVE\ndata: ""\n\n\n'                  │      │
│ │   chunk_size = None                                                 │      │
│ │      chunker = <httpx._decoders.TextChunker object at 0x7f88d63c90> │      │
│ │      decoder = <httpx._decoders.TextDecoder object at 0x7f89884950> │      │
│ │         self = <Response [200 OK]>                                  │      │
│ │ text_content = 'event: KEEP-ALIVE\ndata: ""\n\n\n'                  │      │
│ ╰─────────────────────────────────────────────────────────────────────╯      │
│                                                                              │
│ /root/.cache/pypoetry/virtualenvs/aiohomeconnect-oJCSMThX-py3.11/lib/python3 │
│ .11/site-packages/httpx/_models.py:997 in aiter_bytes                        │
│                                                                              │
│    994 │   │   │   decoder = self._get_content_decoder()                     │
│    995 │   │   │   chunker = ByteChunker(chunk_size=chunk_size)              │
│    996 │   │   │   with request_context(request=self._request):              │
│ ❱  997 │   │   │   │   async for raw_bytes in self.aiter_raw():              │
│    998 │   │   │   │   │   decoded = decoder.decode(raw_bytes)               │
│    999 │   │   │   │   │   for chunk in chunker.decode(decoded):             │
│   1000 │   │   │   │   │   │   yield chunk                                   │
│                                                                              │
│ ╭─────────────────────────────── locals ────────────────────────────────╮    │
│ │      chunk = b'event: KEEP-ALIVE\ndata: ""\n\n\n'                     │    │
│ │ chunk_size = None                                                     │    │
│ │    chunker = <httpx._decoders.ByteChunker object at 0x7f88deffd0>     │    │
│ │    decoded = b'event: KEEP-ALIVE\ndata: ""\n\n\n'                     │    │
│ │    decoder = <httpx._decoders.IdentityDecoder object at 0x7f88defd10> │    │
│ │  raw_bytes = b'event: KEEP-ALIVE\ndata: ""\n\n\n'                     │    │
│ │       self = <Response [200 OK]>                                      │    │
│ ╰───────────────────────────────────────────────────────────────────────╯    │
│                                                                              │
│ /root/.cache/pypoetry/virtualenvs/aiohomeconnect-oJCSMThX-py3.11/lib/python3 │
│ .11/site-packages/httpx/_models.py:1055 in aiter_raw                         │
│                                                                              │
│   1052 │   │   chunker = ByteChunker(chunk_size=chunk_size)                  │
│   1053 │   │                                                                 │
│   1054 │   │   with request_context(request=self._request):                  │
│ ❱ 1055 │   │   │   async for raw_stream_bytes in self.stream:                │
│   1056 │   │   │   │   self._num_bytes_downloaded += len(raw_stream_bytes)   │
│   1057 │   │   │   │   for chunk in chunker.decode(raw_stream_bytes):        │
│   1058 │   │   │   │   │   yield chunk                                       │
│                                                                              │
│ ╭──────────────────────────────── locals ─────────────────────────────────╮  │
│ │            chunk = b'event: KEEP-ALIVE\ndata: ""\n\n\n'                 │  │
│ │       chunk_size = None                                                 │  │
│ │          chunker = <httpx._decoders.ByteChunker object at 0x7f88defe50> │  │
│ │ raw_stream_bytes = b'event: KEEP-ALIVE\ndata: ""\n\n\n'                 │  │
│ │             self = <Response [200 OK]>                                  │  │
│ ╰─────────────────────────────────────────────────────────────────────────╯  │
│                                                                              │
│ /root/.cache/pypoetry/virtualenvs/aiohomeconnect-oJCSMThX-py3.11/lib/python3 │
│ .11/site-packages/httpx/_client.py:176 in __aiter__                          │
│                                                                              │
│    173 │   │   self._start = start                                           │
│    174 │                                                                     │
│    175 │   async def __aiter__(self) -> typing.AsyncIterator[bytes]:         │
│ ❱  176 │   │   async for chunk in self._stream:                              │
│    177 │   │   │   yield chunk                                               │
│    178 │                                                                     │
│    179 │   async def aclose(self) -> None:                                   │
│                                                                              │
│ ╭──────────────────────────── locals ─────────────────────────────╮          │
│ │ chunk = b'event: KEEP-ALIVE\ndata: ""\n\n\n'                    │          │
│ │  self = <httpx._client.BoundAsyncStream object at 0x7f88d4c5d0> │          │
│ ╰─────────────────────────────────────────────────────────────────╯          │
│                                                                              │
│ /root/.cache/pypoetry/virtualenvs/aiohomeconnect-oJCSMThX-py3.11/lib/python3 │
│ .11/site-packages/httpx/_transports/default.py:270 in __aiter__              │
│                                                                              │
│   267 │   │   self._httpcore_stream = httpcore_stream                        │
│   268 │                                                                      │
│   269 │   async def __aiter__(self) -> typing.AsyncIterator[bytes]:          │
│ ❱ 270 │   │   with map_httpcore_exceptions():                                │
│   271 │   │   │   async for part in self._httpcore_stream:                   │
│   272 │   │   │   │   yield part                                             │
│   273                                                                        │
│                                                                              │
│ ╭───────────────────────────────── locals ─────────────────────────────────╮ │
│ │ part = b'event: KEEP-ALIVE\ndata: ""\n\n\n'                              │ │
│ │ self = <httpx._transports.default.AsyncResponseStream object at          │ │
│ │        0x7f88d4fa90>                                                     │ │
│ ╰──────────────────────────────────────────────────────────────────────────╯ │
│                                                                              │
│ /usr/lib/python3.11/contextlib.py:155 in __exit__                            │
│                                                                              │
│   152 │   │   │   │   # tell if we get the same exception back               │
│   153 │   │   │   │   value = typ()                                          │
│   154 │   │   │   try:                                                       │
│ ❱ 155 │   │   │   │   self.gen.throw(typ, value, traceback)                  │
│   156 │   │   │   except StopIteration as exc:                               │
│   157 │   │   │   │   # Suppress StopIteration *unless* it's the same except │
│   158 │   │   │   │   # was passed to throw().  This prevents a StopIteratio │
│                                                                              │
│ ╭───────────────────────────────── locals ─────────────────────────────────╮ │
│ │      self = <contextlib._GeneratorContextManager object at 0x7f88defe90> │ │
│ │ traceback = <traceback object at 0x7f88dbd080>                           │ │
│ │     value = RemoteProtocolError(RemoteProtocolError('peer closed         │ │
│ │             connection without sending complete message body (incomplete │ │
│ │             chunked read)'))                                             │ │
│ ╰──────────────────────────────────────────────────────────────────────────╯ │
│                                                                              │
│ /root/.cache/pypoetry/virtualenvs/aiohomeconnect-oJCSMThX-py3.11/lib/python3 │
│ .11/site-packages/httpx/_transports/default.py:118 in                        │
│ map_httpcore_exceptions                                                      │
│                                                                              │
│   115 │   │   │   raise                                                      │
│   116 │   │                                                                  │
│   117 │   │   message = str(exc)                                             │
│ ❱ 118 │   │   raise mapped_exc(message) from exc                             │
│   119                                                                        │
│   120                                                                        │
│   121 class ResponseStream(SyncByteStream):                                  │
│                                                                              │
│ ╭───────────────────────────────── locals ─────────────────────────────────╮ │
│ │ message = 'peer closed connection without sending complete message body  │ │
│ │           (incomplete chunked'+6                                         │ │
│ ╰──────────────────────────────────────────────────────────────────────────╯ │
╰──────────────────────────────────────────────────────────────────────────────╯
RemoteProtocolError: peer closed connection without sending complete message 
body (incomplete chunked read)
```

</details>

Reply:
I tested disconnection also, and it rises a `ReadTimeout` exception